File: python_code/56_合并区间.py
'''
56. 合并区间
给出一个区间的集合，请合并所有重叠的区间。

示例 1:

输入: [[1,3],[2,6],[8,10],[15,18]]
输出: [[1,6],[8,10],[15,18]]
解释: 区间 [1,3] 和 [2,6] 重叠, 将它们合并为 [1,6].
示例 2:

输入: [[1,4],[4,5]]
输出: [[1,5]]
解释: 区间 [1,4] 和 [4,5] 可被视为重叠区间。
'''

class Solution:
    def merge(self, intervals:List[List[int]])->List[List[int]]:
        intervals.sort(key=lambda x:x[0])
        merged= []
        for interval in intervals:
            if not merged or merged[-1][1]<intervals[0]:
                merged.append(interval)
            else:
                merged[-1][1] = max(merged[-1][1],interval[1])
        return merged

File: C++_code/56_合并区间.cpp
/*
56. 合并区间
给出一个区间的集合，请合并所有重叠的区间。

示例 1:

输入: [[1,3],[2,6],[8,10],[15,18]]
输出: [[1,6],[8,10],[15,18]]
解释: 区间 [1,3] 和 [2,6] 重叠, 将它们合并为 [1,6].
示例 2:

输入: [[1,4],[4,5]]
输出: [[1,5]]
解释: 区间 [1,4] 和 [4,5] 可被视为重叠区间。
*/

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        if(intervals.size()==0) return {};

        sort(intervals.begin(),intervals.end());
        vector<vector<int>> merge;
        for(int i=0;i<intervals.size();i++)
        {
            if(!merge.size()||merge.back()[1]<intervals[i][0])
                merge.push_back({intervals[i][0],intervals[i][1]});
            else
                merge.back()[1] = max(merge.back()[1],intervals[i][1]);
        }
        return merge;
    }
};